lib/THC/THCTensorIndex.cu

```diff
@@ -94,6 +95,124 @@ __global__ void indexCopyLargeIndex(TensorInfo<T, IndexType> dst,
   }
 }
 
+template <typename T, size_t n>
+struct AtomicAddIntegerImpl;
+
+template<typename T>
+struct AtomicAddIntegerImpl<T, 1> {
+  __device__ void operator()(T *address, T val) {
+    unsigned int * address_as_ui =
+        (unsigned int *) (address - ((size_t)address & 3));
+    unsigned int old = *address_as_ui, sum, newval, assumed;
+    unsigned int selectors[] = {0x3214, 0x3240, 0x3410, 0x4210};
+    unsigned int sel = selectors[(size_t)address & 3];
+
+    do {
+      assumed = old;
+      sum = val +  (T)__byte_perm(old, 0, (size_t)address & 3 | 0x4440);
```

Comment:
nevermind, i guess there isn't overflow because you're selecting again below to re-permute, but i didn't stare at the selection indices long enough to determine if it was correct. I think just shifting + masking the 8 or 16 bit word from the 32 bit word that you want and then putting it back is easier to follow, and then you aren't using an array of selector values either based on offset (which might go into local memory or involve indexing in constant memory)